runtime/eventlog.c

```diff
@@ -258,13 +355,14 @@ void caml_eventlog_init()
 #endif
 
   setup_eventlog_file();
+  setup_metadata_file();
   setup_evbuf();
 
   atexit(&teardown_eventlog);
 }
 
 static void post_event(ev_gc_phase phase, ev_gc_counter counter_kind,
-                       uint8_t bucket, uint64_t count, ev_type ty)
+                       uint8_t bucket, uint64_t count, ev_user_type span_type, ev_type ty)
```

Comment:
Rather than `ev_user_type` and `ev_type`, and having `ev_user_type` to be 0 for GC events. Here is an alternate proposal that I think would be cleaner and more compact. 

Introduce `EV_USER_BEGIN` and `EV_USER_END` in the [`ev_type` enum](https://github.com/gshag/ocaml/pull/1/files#diff-d0a4a1bdecf3c22bef07c0cfbc49d9b0R26). Rename the `EV_ENTRY` and `EV_EXIT` to `EV_GC_BEGIN` and `EV_GC_END`. Move `MIN_AVAILABLE_EVENT_ID` definition to the last entry in the [`ev_gc_phase` enum](https://github.com/ocaml/ocaml/blob/trunk/runtime/caml/eventlog.h#L60), so that the value of `MIN_AVAILABLE_EVENT_ID` stays in sync. Now rather than have both `ev_gc_phase` and `span_type`, one of which is not used based on whether the event is a GC event or User event, you could just encode the user event identified in the `gc_phase` field. On that note, it would be useful to rename `ev_gc_phase` to `ev_span_type` to better reflect their intended use. You can discriminate a GC span from a user span based on `ev_type` and the `ev_span_type` reflects the appropriate span type.

Reply:
Generally, the existing implementation can also be improved since the allocation counter-event and the span events are mixed together, and there are unused fields in the emitted event when span events are emitted?

Reply:
I think I understood the intent of the comment and have made some changes. They're not exactly the same as the suggestions and here's how and why they differ:

- I have combined `ev_user_type` and `ev_gc_phase` to become `ev_span_type`. This is the same as `ev_gc_phase` with the additions of `EV_USER_BEGIN` and `EV_USER_END`.
- I have added `MIN_AVAILABLE_EVENT_ID` as the last entry in `ev_type`.

Difference from the original comment: The suggestion was to combine `ev_user_type` and `ev_type`, while adding `MIN_AVAILABLE_EVENT_ID` to `ev_gc_phase` (which was to be renamed as `ev_span_type`). So in some sense I've mixed the opposite things.

Reasoning: `ev_type` goes in the event header and is used to match the emitted event with the event description in the metadata. With user events we are emitting a new type of event in the metadata stream for every new user event. So the `ev_type` is in some sense just a numbering system for the events, and since GC event types are fixed, it is just a numbering system for user events in particular that begins at `MIN_AVAILABLE_EVENT_ID`. Therefore it made sense to fix `MIN_AVAILABLE_EVENT_ID` according to the fixed entries in `ev_type`. Now we have a way to distinguish between user and GC events by virtue of the `ev_type`, so we can combine `ev_gc_phase` and `ev_user_type` as the comment suggested.

Re: unused fields that are emitted - as per discussion, these unused fields only appear for code unification and are not actually emitted.

I would be interested in discussing the allocation counter-event and the span event mixture problem if that would be useful to sort out!